File: backend/src/services/manual_tags_service.py
"""Manual Tags Service
Service layer for managing user-created manual tags on sentences
"""

import uuid
from datetime import datetime
from typing import Dict, List, Optional

from ..models.manual_tags import (
    ManualTag, 
    ManualTagType, 
    CreateTagRequest, 
    UpdateTagRequest
)


class ManualTagsService:
    """Service for managing manual tags"""
    
    def __init__(self):
        # In-memory storage for now (could be replaced with database later)
        self._tags: Dict[str, ManualTag] = {}
        self._tags_by_analysis: Dict[str, List[str]] = {}  # analysis_id -> tag_ids
    
    def create_tag(self, request: CreateTagRequest) -> ManualTag:
        """Create a new manual tag"""
        tag_id = str(uuid.uuid4())
        
        tag = ManualTag(
            id=tag_id,
            tag_type=request.tag_type,
            sentence_index=request.sentence_index,
            sentence_text=request.sentence_text,
            analysis_id=request.analysis_id,
            user_notes=request.user_notes,
            created_at=datetime.now(),
            updated_at=datetime.now()
        )
        
        # Store the tag
        self._tags[tag_id] = tag
        
        # Index by analysis_id
        if request.analysis_id not in self._tags_by_analysis:
            self._tags_by_analysis[request.analysis_id] = []
        self._tags_by_analysis[request.analysis_id].append(tag_id)
        
        return tag
    
    def get_tag(self, tag_id: str) -> Optional[ManualTag]:
        """Get a specific tag by ID"""
        return self._tags.get(tag_id)
    
    def get_tags_for_analysis(self, analysis_id: str) -> List[ManualTag]:
        """Get all tags for a specific analysis session"""
        tag_ids = self._tags_by_analysis.get(analysis_id, [])
        return [self._tags[tag_id] for tag_id in tag_ids if tag_id in self._tags]
    
    def get_tags_for_sentence(self, analysis_id: str, sentence_index: int) -> List[ManualTag]:
        """Get all tags for a specific sentence in an analysis"""
        analysis_tags = self.get_tags_for_analysis(analysis_id)
        return [tag for tag in analysis_tags if tag.sentence_index == sentence_index]
    
    def update_tag(self, tag_id: str, request: UpdateTagRequest) -> Optional[ManualTag]:
        """Update an existing tag"""
        tag = self._tags.get(tag_id)
        if not tag:
            return None
        
        # Update fields if provided
        if request.tag_type is not None:
            tag.tag_type = request.tag_type
        if request.user_notes is not None:
            tag.user_notes = request.user_notes
        
        tag.updated_at = datetime.now()
        
        return tag
    
    def delete_tag(self, tag_id: str) -> bool:
        """Delete a tag"""
        tag = self._tags.get(tag_id)
        if not tag:
            return False
        
        # Remove from main storage
        del self._tags[tag_id]
        
        # Remove from analysis index
        analysis_id = tag.analysis_id
        if analysis_id in self._tags_by_analysis:
            if tag_id in self._tags_by_analysis[analysis_id]:
                self._tags_by_analysis[analysis_id].remove(tag_id)
        
        return True
    
    def delete_tags_for_analysis(self, analysis_id: str) -> int:
        """Delete all tags for an analysis session"""
        tag_ids = self._tags_by_analysis.get(analysis_id, []).copy()
        deleted_count = 0
        
        for tag_id in tag_ids:
            if self.delete_tag(tag_id):
                deleted_count += 1
        
        # Clean up the analysis index
        if analysis_id in self._tags_by_analysis:
            del self._tags_by_analysis[analysis_id]
        
        return deleted_count
    
    def get_available_tag_types(self) -> Dict[str, str]:
        """Get available manual tag types with descriptions"""
        # Official descriptions from the system
        tag_descriptions = {
            ManualTagType.AS_PLUS.value: "Alteração de Sentido",
            ManualTagType.DL_PLUS.value: "Reorganização Posicional",
            ManualTagType.EXP_PLUS.value: "Explicitação e Detalhamento",
            ManualTagType.IN_PLUS.value: "Manejo de Inserções",
            ManualTagType.MOD_PLUS.value: "Reinterpretação Perspectiva",
            ManualTagType.MT_PLUS.value: "Otimização de Títulos",
            ManualTagType.OM_PLUS.value: "Supressão Seletiva",
            ManualTagType.PRO_PLUS.value: "Desvio Semântico",
            ManualTagType.RF_PLUS.value: "Reescrita Global",
            ManualTagType.RD_PLUS.value: "Estruturação de Conteúdo e Fluxo",
            ManualTagType.RP_PLUS.value: "Fragmentação Sintática",
            ManualTagType.SL_PLUS.value: "Adequação de Vocabulário",
            ManualTagType.TA_PLUS.value: "Clareza Referencial",
            ManualTagType.MV_PLUS.value: "Alteração da Voz Verbal",
        }
        return tag_descriptions
    
    def get_stats(self, analysis_id: Optional[str] = None) -> Dict[str, any]:
        """Get statistics about tags"""
        if analysis_id:
            tags = self.get_tags_for_analysis(analysis_id)
        else:
            tags = list(self._tags.values())
        
        tag_type_counts = {}
        for tag in tags:
            tag_type = tag.tag_type.value
            tag_type_counts[tag_type] = tag_type_counts.get(tag_type, 0) + 1
        
        return {
            "total_tags": len(tags),
            "tag_type_distribution": tag_type_counts,
            "analysis_sessions": len(self._tags_by_analysis) if not analysis_id else 1
        }


# Global service instance
manual_tags_service = ManualTagsService()

File: frontend/src/services/manualTagsService.js
/**
 * Manual Tags Service
 * Service for managing user-created manual tags on sentences
 */

import { manualTagsAPI } from './api';

/**
 * Service class for manual tags operations
 */
export class ManualTagsService {
  
  /**
   * Create a new manual tag for a sentence
   * @param {Object} tagData - Tag creation data
   * @param {string} tagData.tagType - Tag type (e.g., 'AS+', 'SL+')
   * @param {number} tagData.sentenceIndex - Index of the sentence to tag
   * @param {string} tagData.sentenceText - Text of the sentence being tagged
   * @param {string} tagData.analysisId - ID of the analysis session
   * @param {string} [tagData.userNotes] - Optional user notes
   * @returns {Promise<Object>} Created tag data
   */
  static async createTag(tagData) {
    try {
      console.log('Creating manual tag:', tagData);
      
      const request = {
        tag_type: tagData.tagType,
        sentence_index: tagData.sentenceIndex,
        sentence_text: tagData.sentenceText,
        analysis_id: tagData.analysisId,
        user_notes: tagData.userNotes || null
      };
      
      const response = await manualTagsAPI.create(request);
      
      if (response.data.success) {
        console.log('Manual tag created successfully:', response.data.tag);
        return {
          success: true,
          tag: response.data.tag,
          message: response.data.message || 'Tag criada com sucesso!'
        };
      } else {
        return {
          success: false,
          message: response.data.message || 'Erro ao criar tag'
        };
      }
      
    } catch (error) {
      console.error('Failed to create manual tag:', error);
      throw new Error(
        error.response?.data?.detail || 
        error.response?.data?.message ||
        'Erro ao criar tag manual'
      );
    }
  }

  /**
   * Get all manual tags for an analysis session
   * @param {string} analysisId - Analysis session ID
   * @returns {Promise<Array>} Array of manual tags
   */
  static async getTagsForAnalysis(analysisId) {
    try {
      console.log('Getting tags for analysis:', analysisId);
      
      const response = await manualTagsAPI.getForAnalysis(analysisId);
      
      if (response.data.success) {
        console.log(`Retrieved ${response.data.tags.length} tags for analysis ${analysisId}`);
        return response.data.tags;
      } else {
        console.warn('Failed to get tags:', response.data.message);
        return [];
      }
      
    } catch (error) {
      console.error('Failed to get tags for analysis:', error);
      // Return empty array on error instead of throwing, as tags are optional
      return [];
    }
  }

  /**
   * Get manual tags for a specific sentence
   * @param {string} analysisId - Analysis session ID
   * @param {number} sentenceIndex - Sentence index
   * @returns {Promise<Array>} Array of manual tags for the sentence
   */
  static async getTagsForSentence(analysisId, sentenceIndex) {
    try {
      console.log('Getting tags for sentence:', analysisId, sentenceIndex);
      
      const response = await manualTagsAPI.getForSentence(analysisId, sentenceIndex);
      
      if (response.data.success) {
        return response.data.tags;
      } else {
        return [];
      }
      
    } catch (error) {
      console.error('Failed to get tags for sentence:', error);
      return [];
    }
  }

  /**
   * Update an existing manual tag
   * @param {string} tagId - Tag ID to update
   * @param {Object} updates - Updates to apply
   * @param {string} [updates.tagType] - New tag type
   * @param {string} [updates.userNotes] - New user notes
   * @returns {Promise<Object>} Updated tag data
   */
  static async updateTag(tagId, updates) {
    try {
      console.log('Updating manual tag:', tagId, updates);
      
      const request = {
        tag_type: updates.tagType || null,
        user_notes: updates.userNotes || null
      };
      
      const response = await manualTagsAPI.update(tagId, request);
      
      if (response.data.success) {
        console.log('Manual tag updated successfully:', response.data.tag);
        return {
          success: true,
          tag: response.data.tag,
          message: response.data.message || 'Tag atualizada com sucesso!'
        };
      } else {
        return {
          success: false,
          message: response.data.message || 'Erro ao atualizar tag'
        };
      }
      
    } catch (error) {
      console.error('Failed to update manual tag:', error);
      throw new Error(
        error.response?.data?.detail || 
        error.response?.data?.message ||
        'Erro ao atualizar tag manual'
      );
    }
  }

  /**
   * Delete a manual tag
   * @param {string} tagId - Tag ID to delete
   * @returns {Promise<Object>} Deletion result
   */
  static async deleteTag(tagId) {
    try {
      console.log('Deleting manual tag:', tagId);
      
      const response = await manualTagsAPI.delete(tagId);
      
      if (response.data.success) {
        console.log('Manual tag deleted successfully');
        return {
          success: true,
          message: response.data.message || 'Tag excluída com sucesso!',
          deletedTagId: response.data.deleted_tag_id
        };
      } else {
        return {
          success: false,
          message: response.data.message || 'Erro ao excluir tag'
        };
      }
      
    } catch (error) {
      console.error('Failed to delete manual tag:', error);
      throw new Error(
        error.response?.data?.detail || 
        error.response?.data?.message ||
        'Erro ao excluir tag manual'
      );
    }
  }

  /**
   * Get available tag types with descriptions
   * @returns {Promise<Object>} Tag types mapping
   */
  static async getAvailableTagTypes() {
    try {
      const response = await manualTagsAPI.getAvailableTypes();
      return response.data;
    } catch (error) {
      console.error('Failed to get available tag types:', error);
      // Return default tag types if API fails
      return {
        'AS+': 'Alteração de Sentido',
        'DL+': 'Reorganização Posicional',
        'EXP+': 'Explicitação e Detalhamento',
        'IN+': 'Manejo de Inserções',
        'MOD+': 'Reinterpretação Perspectiva',
        'MT+': 'Otimização de Títulos',
        'OM+': 'Supressão Seletiva',
        'PRO+': 'Desvio Semântico',
        'RF+': 'Reescrita Global',
        'RD+': 'Estruturação de Conteúdo e Fluxo',
        'RP+': 'Fragmentação Sintática',
        'SL+': 'Adequação de Vocabulário',
        'TA+': 'Clareza Referencial',
        'MV+': 'Alteração da Voz Verbal'
      };
    }
  }

  /**
   * Get statistics for manual tags in an analysis
   * @param {string} analysisId - Analysis session ID
   * @returns {Promise<Object>} Tag statistics
   */
  static async getTagStats(analysisId) {
    try {
      const response = await manualTagsAPI.getStats(analysisId);
      if (response.data.success) {
        return response.data.stats;
      } else {
        return {
          total_tags: 0,
          tag_type_distribution: {},
          analysis_sessions: 0
        };
      }
    } catch (error) {
      console.error('Failed to get tag stats:', error);
      return {
        total_tags: 0,
        tag_type_distribution: {},
        analysis_sessions: 0
      };
    }
  }

  /**
   * Delete all manual tags for an analysis session
   * @param {string} analysisId - Analysis session ID
   * @returns {Promise<Object>} Deletion result
   */
  static async deleteAllTagsForAnalysis(analysisId) {
    try {
      console.log('Deleting all tags for analysis:', analysisId);
      
      const response = await manualTagsAPI.deleteAllForAnalysis(analysisId);
      
      if (response.data.success) {
        console.log(`Deleted ${response.data.deleted_count} tags for analysis ${analysisId}`);
        return {
          success: true,
          message: response.data.message,
          deletedCount: response.data.deleted_count
        };
      } else {
        return {
          success: false,
          message: 'Erro ao excluir tags da análise'
        };
      }
      
    } catch (error) {
      console.error('Failed to delete tags for analysis:', error);
      throw new Error(
        error.response?.data?.detail || 
        error.response?.data?.message ||
        'Erro ao excluir tags da análise'
      );
    }
  }

  /**
   * Check if a tag already exists for a sentence
   * @param {Array} existingTags - Array of existing tags
   * @param {number} sentenceIndex - Sentence index to check
   * @param {string} tagType - Tag type to check
   * @returns {boolean} True if tag exists
   */
  static tagExistsForSentence(existingTags, sentenceIndex, tagType) {
    return existingTags.some(tag => 
      tag.sentence_index === sentenceIndex && 
      tag.tag_type === tagType
    );
  }

  /**
   * Format tag for display in UI
   * @param {Object} tag - Raw tag object from API
   * @returns {Object} Formatted tag object
   */
  static formatTagForUI(tag) {
    return {
      id: tag.id,
      code: tag.tag_type,
      name: tag.tag_type, // For backward compatibility
      fullName: tag.tag_type, // For backward compatibility
      confidence: tag.confidence || 1.0,
      evidence: tag.evidence || ['Manually added by human validator'],
      sentenceIndex: tag.sentence_index,
      sentenceText: tag.sentence_text,
      analysisId: tag.analysis_id,
      userNotes: tag.user_notes,
      createdAt: tag.created_at,
      updatedAt: tag.updated_at,
      isManual: true // Flag to distinguish from automatic tags
    };
  }

  /**
   * Convert UI format back to API format
   * @param {Object} uiTag - Tag in UI format
   * @returns {Object} Tag in API format
   */
  static formatTagForAPI(uiTag) {
    return {
      tag_type: uiTag.code || uiTag.name,
      sentence_index: uiTag.sentenceIndex,
      sentence_text: uiTag.sentenceText,
      analysis_id: uiTag.analysisId,
      user_notes: uiTag.userNotes || null
    };
  }
}

export default ManualTagsService;